Scroll anchoring: naive implementation of selectAnchorElement(), implement updateScrollPosition(), hookup to layout

nmoucht opened the pull request:
#### 6639646de774d4333e28e4288a36c5627ac6388f
<pre>
Scroll anchoring: naive implementation of selectAnchorElement(), implement updateScrollPosition(), hookup to layout
<a href="https://bugs.webkit.org/show_bug.cgi?id=243702">https://bugs.webkit.org/show_bug.cgi?id=243702</a>
&lt;rdar://98695739&gt;

Reviewed by Simon Fraser.

For selectAnchorElement(), select the deepest child element that is within the viewport bounds of the
frame view, or the first fully contined by the viewport bounds. For updateScrollPosition()
calculate the change in offset and set the scroll position of the frame view if necessary.
Finally, in FrameViewLayoutContext::performLayout(), hookup calls to selectAnchorElement()
and updateScrollPosition() in the appropriate area.

* Source/WebCore/page/FrameView.cpp:
(WebCore::FrameView::willDoLayout):
(WebCore::FrameView::performPostLayoutTasks):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::doAfterUpdateRendering):
* Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
(WebCore::ScrollAnchoringController::invalidateAnchorElement):
(WebCore::ScrollAnchoringController::computeOffset):
(WebCore::ScrollAnchoringController::selectAnchorElement):
(WebCore::ScrollAnchoringController::updateScrollPosition):
* Source/WebCore/page/scrolling/ScrollAnchoringController.h:

Canonical link: <a href="https://commits.webkit.org/268267@main">https://commits.webkit.org/268267@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/25e7962b9d6843809d19f3198eeece0267b907c7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/19189 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/19603 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/20201 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/21078 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17960 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/19404 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/22875 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/19731 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/19639 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/19410 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/19484 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16689 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/21953 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/16679 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/17475 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/23839 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17732 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/17649 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/21793 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/18241 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/15452 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/17364 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/17312 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4587 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21724 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/18077 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
```diff
@@ -25,30 +25,68 @@
+    tree.append(rootRenderer->firstChildBox());
+    while (!tree.isEmpty()) {
+        auto box = tree.last();
+        tree.removeLast();
```
you could just use takeLast() instead

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/FrameView.cpp:
```diff
@@ -1271,7 +1271,10 @@ void FrameView::willDoLayout(WeakPtr<RenderElement> layoutRoot)
-        
+    
+    if (m_scrollAnchoringController)
+        m_scrollAnchoringController->selectAnchorBox();
```
Layout might be dirty at this point (style will have changed, but the render tree is still in the old state). So it's not a good time to query the render tree about geometry. I think you need to run `selectAnchorBox()` after each scroll and layout, in a way that doesn't hurt performance.

nmoucht replied:
I'm a bit confused by the suggestion. Are you suggesting somewhere after layout (like `performPostLayoutTasks()`)? Also the implementation right now is invalidating the anchor node after a scroll, so I don't think its necessary to calculate the anchor after each scroll (the spec allows this implementation for lazy anchor node selection: 

> Conceptually, a new anchor node is computed for every scrolling box whenever the scroll position of any scrolling box changes. (As a performance optimization, the implementation may wait until the anchor node is needed before computing it.)

nmoucht replied:
Moved to runScrollSteps/scrollPositionChanged to avoid this.

smfr commented on Source/WebCore/page/FrameView.cpp:
```diff
@@ -3572,6 +3575,8 @@ void FrameView::performPostLayoutTasks()
 
     scrollToAnchor();
+    if (m_scrollAnchoringController)
+        m_scrollAnchoringController->updateScrollPosition();
```
I know scrollToAnchor() is here, but this can run multiple times between rendering updates because of forced layouts. A better time to do this is inside `document.runScrollSteps();` (and I'd expect the spec to say that?).

nmoucht replied:
I tried adding it to document.runScrollSteps() (I think you are right that it should go here), but it was causing the tests to fail even though I was making the call to `m_frameView.setScrollPosition()` with the correct offset. My assumption is that we need to trigger a render update after the scroll right? Calling `page()->scheduleRenderingUpdate()` didn't seem to affect this.

smfr replied:
Did the tests fail because they were written in a way that expects some kind of synchronous anchoring behavior?

I would expect that the triggering of scroll events already schedules a rendering update after a scroll.

nmoucht replied:
I added a call to updateScrollPositionForAnchorElement in Document::updateLayoutIgnorePendingStylesheets to ensure calls to scrollTop/etc would get the current state including any adjustments due to scroll anchoring.

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
```diff
@@ -25,30 +25,68 @@
-    // Firefox appears to only iterate through frame tree under the current frame (unclear if
-    // looking through frame tree is sufficient)
-    // unclear to me where to call this from 
+    return candidate->location() - m_frameView.rectForFixedPositionLayout().location();
```
`RenderBox::location()` is relative to its containing block, but you really need "absolute" (i.e. document-relative) coordinates here, or "client" (relative to viewport) coordinates. If you use `client` coordinates, check that they are relative to the `layoutViewport` (they may deviate on zooming).

Also, instead of `rectForFixedPositionLayout`, use `layoutViewport`.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
```diff
@@ -25,30 +25,68 @@
+    return candidate->location() - m_frameView.rectForFixedPositionLayout().location();
+}
+
+void ScrollAnchoringController::selectAnchorBox()
```
"select" is a bit overloaded (it might imply Selection). Maybe call this `chooseAnchorBox` or something.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
```diff
@@ -25,30 +25,68 @@
+    // TODO: fully implement anchor node selection algorithm in spec
+    // (specifically excluded subtrees and priority nodes)
+    auto scrollPosition = frameView().scrollPosition();
+    if (m_anchorBox || scrollPosition == IntPoint())
```
Is this `scrollPosition == IntPoint()` in the spec? (Better as `scrollPosition.isZero()`)

You need to do some testing with RTL documents where the initial scroll position is non-zero, and the scroll origin is non-zero (and write WPTs if there aren't any).

nmoucht replied:
It is one of the suppression triggers: The scroll offset of the scrollable element being zero. https://drafts.csswg.org/css-scroll-anchoring/#suppression-triggers.

nmoucht replied:
Could you explain your second point a bit more (I'm a bit confused as to why RTL is relevant here)?

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
```diff
@@ -25,30 +25,68 @@
+    auto& document = *m_frameView.frame().document();
+    auto* documentElement = document.documentElement();
+    RenderBox* rootRenderer = documentElement ? documentElement->renderBox() : nullptr;
+        RenderBox* candidateAnchor = nullptr;
```
Bad indentation.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
```diff
@@ -25,30 +25,68 @@
+    auto* documentElement = document.documentElement();
+    RenderBox* rootRenderer = documentElement ? documentElement->renderBox() : nullptr;
+        RenderBox* candidateAnchor = nullptr;
+    auto viewableRect = m_frameView.rectForFixedPositionLayout();
```
layoutViewport ?

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
```diff
@@ -25,30 +25,68 @@
+        RenderBox* candidateAnchor = nullptr;
+    auto viewableRect = m_frameView.rectForFixedPositionLayout();
+    
+    Vector<RenderBox*> tree;
```
This looks like a list, not a tree.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
```diff
@@ -25,30 +25,68 @@
+        tree.removeLast();
+        if (box->style().overflowAnchor() == OverflowAnchor::None)
+            continue;
+        auto boxRect = box->frameRect();
```
frameRect is local to the box. You need to convert to `absolute` coordinates if you're comparing with the layoutViewport (I think; you should check this, and test with zooming).

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
```diff
@@ -25,30 +25,68 @@
+
+void ScrollAnchoringController::selectAnchorBox()
+{
+    // TODO: fully implement anchor node selection algorithm in spec
```
It would be good to have a link to the spec here

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
```diff
@@ -25,30 +25,68 @@
+            break;
+        if (!is<RenderBox>(box->firstChild()))
+            continue;
+        for (auto child = box->firstChildBox(); child; child = child->nextSiblingBox()) {
```
The spec talks in terms of DOM nodes, not boxes, so does this algorithm need to run on Nodes/Elements?

nmoucht replied:
The spec does make mention of boxes: 

> An anchor node can be any [box](https://drafts.csswg.org/css-display-3/#box) except one for a non-[atomic inline](https://drafts.csswg.org/css-display-3/#atomic-inline)

I believe the Google implementation also runs their selection algorithm on the box tree. Functionally it should be the same right?

smfr replied:
The box tree can differ from the DOM tree because of `display:contents`, `display:none`, fragmentation and many other reasons. Some of these will be observable (e.g. maybe think about scroll anchoring where the anchorBox is something split across 2 columns in multicol).

nmoucht replied:
Changed to DOM tree walk.

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
```diff
@@ -25,30 +25,68 @@
-    m_anchorElement = nullptr;
-    m_lastPositionForAnchorElement = FloatPoint();
+    m_anchorBox = nullptr;
+    m_lastPositionForAnchorBox = FloatSize();
```
` = { }`

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
```diff
@@ -25,30 +25,68 @@
+        return;
+    FloatSize adjustment = computeOffset(m_anchorBox.get()) - m_lastPositionForAnchorBox;
+    if (!adjustment.isZero())
+        m_frameView.setScrollPosition(frameView().scrollPosition() + IntPoint(adjustment.width(), adjustment.height()));
```
Please add logging with `LOG(Scroll, ...)` here so we know when scroll anchoring is changing the scroll position.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.h:
```diff
@@ -30,23 +30,25 @@
-    WeakPtr<Element> m_anchorElement;
-    FloatPoint m_lastPositionForAnchorElement;
+    WeakPtr<RenderBox> m_anchorBox;
+    FloatSize m_lastPositionForAnchorBox;
```
Generally we'd use `FloatPoint` for a position, and `FloatSize` for an offset (or delta).

nmoucht replied:
This is technically an offset, so changed the name to reflect that.

whsieh commented on Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
```diff
@@ -25,30 +25,68 @@
 }
 
-void ScrollAnchoringController::selectAnchorElement()
+FloatSize ScrollAnchoringController::computeOffset(RenderBox* candidate)
```
Nit - perhaps this helper should take a `const RenderBox&` since it expects a non-null `RenderBox` (and doesn't mutate it in any way)

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/LocalFrameView.h:
```diff
@@ -735,6 +735,11 @@ class LocalFrameView final : public FrameView {
     Color scrollbarTrackColorStyle() const final;
     ScrollbarGutter scrollbarGutterStyle() const final;
     ScrollbarWidth scrollbarWidthStyle() const final;
+    void updateAnchorElement() final;
```
In FrameView here we should refer to the "ScrollAnchoringElement", so it's easily distinguished from the element referenced by a `#` in the URL which is also called an "anchor element".

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -4609,6 +4612,8 @@ void Document::runScrollSteps()
         }
         if (scrollAnimationsInProgress)
             page()->scheduleRenderingUpdate({ RenderingUpdateStep::Scroll });
+        frameView->updateAnchorElement();
```
Add a blank line above.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/LocalFrameView.h:
```diff
@@ -1017,6 +1022,7 @@ class LocalFrameView final : public FrameView {
     ScrollPinningBehavior m_scrollPinningBehavior { ScrollPinningBehavior::DoNotPin };
     SelectionRevealMode m_selectionRevealModeForFocusedElement { SelectionRevealMode::DoNotReveal };
     std::unique_ptr<ScrollAnchoringController> m_scrollAnchoringController;
+    WeakHashSet<ScrollAnchoringController> m_anchorsNeedingUpdate;
```
This is a set of controllers, not a set of anchors. What does it represent - controllers for subframes, or overflow scrollers?

nmoucht replied:
Updated the name to m_scrollAnchoringControllersNeedingUpdate. This represents the set of scroll anchoring controllers needing position updates of the subscrollers for the FrameView (as well as the controller for the frame as well).

smfr replied:
I think it would be better to just store references to ScrollableAreas (we have a set already). And we need to think about how this works with site isolation.

nmoucht replied:
Updated to ScrollableAreaSet (although we need to keep this separate from m_scrollableAreas since not every scrollable area is going to need a scroll position update for each frame).

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.h:
```diff
@@ -41,12 +41,18 @@ class ScrollAnchoringController final {
 
 private:
     LocalFrameView& m_frameView;
+    FloatSize computeOffset(RenderBox& candidate);
```
Use ScrollOffset or ScrollPosition rather than FloatSize.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.h:
```diff
@@ -41,12 +41,18 @@ class ScrollAnchoringController final {
+    FloatSize computeOffset(RenderBox& candidate);
     WeakPtr<Element, WeakPtrImplWithEventTargetData> m_anchorElement;
-    FloatPoint m_lastPositionForAnchorElement;
+    FloatSize m_lastOffsetForAnchorElement;
```
Ditto

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.h:
```diff
@@ -41,12 +41,18 @@ class ScrollAnchoringController final {
-    FloatPoint m_lastPositionForAnchorElement;
+    FloatSize m_lastOffsetForAnchorElement;
+    bool m_midAdjustingScrollPositionForAnchorElement { false };
+    bool m_isQueued;
```
initialize. It's not obvious what "queued" means.

nmoucht replied:
Changed to m_isQueuedForScrollPositionUpdate. This indicates that this controller has a pending scroll position update and is on the set held by the FrameView.

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.h:
```diff
@@ -41,12 +41,18 @@ class ScrollAnchoringController final {
     WeakPtr<Element, WeakPtrImplWithEventTargetData> m_anchorElement;
-    FloatPoint m_lastPositionForAnchorElement;
+    FloatSize m_lastOffsetForAnchorElement;
+    bool m_midAdjustingScrollPositionForAnchorElement { false };
```
not sure what "midAdjusting" is.

nmoucht replied:
Changed to m_midUpdatingScrollPositionForAnchorElement. This indicates that we are currently updating the scroll position in `ScrollAnchoringController::updateScrollPosition` and want to avoid invalidating the anchor element for a scroll caused by a scroll anchoring adjustment.

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
```diff
@@ -25,30 +25,91 @@
+    FloatSize adjustment = computeOffset(*renderBox) - m_lastOffsetForAnchorElement;
+    if (!adjustment.isZero()) {
+        auto newScrollPosition = frameView().scrollPosition() + IntPoint(adjustment.width(), adjustment.height());
+        LOG(Scrolling, "ScrollAnchoringController::updateScrollPosition() scrolling to %d,%d", newScrollPosition.x(), newScrollPosition.y());
```
This should log the frame in which the adjustment is happening. Easier to write with `LOG_WITH_STREAM()`

nmoucht replied:
Fixed.

smfr commented on LayoutTests/platform/ios-wk2/TestExpectations:
```diff
@@ -2291,6 +2291,8 @@ webkit.org/b/243515 svg/compositing/outermost-svg-directly-composited-transforme
 webkit.org/b/243515 svg/compositing/outermost-svg-directly-composited-shape-child.html [ Pass ImageOnlyFailure ]
 webkit.org/b/243515 svg/compositing/outermost-svg-directly-composited-group-child.html [ Pass ImageOnlyFailure ]
 
+webkit.org/b/243702 imported/w3c/web-platform-tests/css/cssom-view/position-sticky-root-scroller-with-scroll-behavior.html [ Failure ]
```
The bug should represent the unfixed reason this tested needed to be marked a failure, rather than tracking the commit that will make it fail (the bug needs to remain open while this test is still marked as a failure).

nmoucht replied:
Made https://bugs.webkit.org/show_bug.cgi?id=261849 to track these tests test failures.

smfr commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -3919,6 +3955,9 @@ void LocalFrameView::scheduleResizeEventIfNeeded()
     }
 
+    invalidateScrollAnchoringElement();
+    updateScrollAnchoringElement();
```
Seems wrong to do this in a function which is about scheduling the resize event. Move this to another resize bottleneck.

nmoucht replied:
This will need to be moved to an area that applies for all ScrollableAreas. This requires more investigation.

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
```diff
@@ -25,30 +25,91 @@
-    // Firefox appears to only iterate through frame tree under the current frame (unclear if
-    // looking through frame tree is sufficient)
-    // unclear to me where to call this from 
+    return ScrollOffset(FloatPoint(candidate.absoluteBoundingBoxRect().location() - m_frameView.layoutViewportRect().location()));
```
Is this right with pinch and command-+ zooming?

smfr replied:
Is it also really a ScrollOffset (i.e. always positive) in RTL documents?

nmoucht replied:
This bit will need to be refactored in https://github.com/WebKit/WebKit/pull/17934 so we can investigate this a bit more once we flesh out the coordinate system we are going to use for subscrollers.

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
```diff
@@ -25,30 +25,91 @@
+
+    Element* candidateAnchor = nullptr;
+    auto layoutViewport = m_frameView.layoutViewportRect();
+    for (auto& element : descendantsOfType<Element>(document->rootNode())) {
```
This is will be wrong when this ScrollAnchoringController exists for an overflow scroll, because you're starting at the document root, and not the root of this scroller.

nmoucht replied:
Yeah this will be addressed in https://github.com/WebKit/WebKit/pull/17934. Added a todo for now.

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
```diff
@@ -25,30 +25,91 @@
+
+void ScrollAnchoringController::updateAnchorElement()
+{
+    if (frameView().scrollPosition().isZero() || m_isQueuedForScrollPositionUpdate)
```
This will be wrong when there's a ScrollAnchoringController per overflow area

nmoucht replied:
Same as above.

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.h:
```diff
@@ -41,12 +41,18 @@ class ScrollAnchoringController final {
-    FloatPoint m_lastPositionForAnchorElement;
+    ScrollOffset m_lastOffsetForAnchorElement;
+    bool m_midUpdatingScrollPositionForAnchorElement { false };
+    bool m_isQueuedForScrollPositionUpdate { false };
```
There's some state aliased between `m_isQueuedForScrollPositionUpdate` and a `ScrollAnchoringController` being a member of the `m_scrollableAreasWithScrollAnchoringControllersNeedingUpdate` set. Try to avoid having both.

nmoucht replied:
I'm not sure there's a great way to get rid of m_isQueuedForScrollPositionUpdate because we have no way of checking if we were queued in `ScrollAnchoringController::adjustScrollPositionForAnchoring` (since the scrollable area is popped by then). However in the future if ScrollAnchoringController::adjustScrollPositionForAnchoring is only called on the queued scrollable areas it might be ok to remove m_isQueuedForScrollPositionUpdate and not check in ScrollAnchoringController::adjustScrollPositionForAnchoring.

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.h:
```diff
@@ -41,12 +41,18 @@ class ScrollAnchoringController final {
     void selectAnchorElement();
+    void chooseAnchorElement();
+    void updateAnchorElement();
+    void printAnchorElement();
```
Unused

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.h:
```diff
@@ -41,12 +41,18 @@ class ScrollAnchoringController final {
     void invalidateAnchorElement();
     void updateScrollPosition();
```
I would call this `adjustScrollPositionForAnchoring()`

nmoucht replied:
Fixed.

nmoucht committed:
Scroll anchoring: naive implementation of selectAnchorElement(), implement updateScrollPosition(), hookup to layout
https://bugs.webkit.org/show_bug.cgi?id=243702
<rdar://98695739>

Reviewed by Simon Fraser.

For selectAnchorElement(), select the deepest child element that is within the viewport bounds of the
frame view, or the first fully contined by the viewport bounds. For updateScrollPosition()
calculate the change in offset and set the scroll position of the frame view if necessary.
Finally, in FrameViewLayoutContext::performLayout(), hookup calls to selectAnchorElement()
and updateScrollPosition() in the appropriate area.

* Source/WebCore/page/FrameView.cpp:
(WebCore::FrameView::willDoLayout):
(WebCore::FrameView::performPostLayoutTasks):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::doAfterUpdateRendering):
* Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
(WebCore::ScrollAnchoringController::invalidateAnchorElement):
(WebCore::ScrollAnchoringController::computeOffset):
(WebCore::ScrollAnchoringController::selectAnchorElement):
(WebCore::ScrollAnchoringController::updateScrollPosition):
* Source/WebCore/page/scrolling/ScrollAnchoringController.h:

Canonical link: https://commits.webkit.org/268267@main

rkirsling commented on Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
```diff
@@ -25,30 +25,96 @@
+    FloatSize adjustment = computeOffset(*renderBox) - m_lastOffsetForAnchorElement;
+    if (!adjustment.isZero()) {
+        auto newScrollPosition = frameView().scrollPosition() + IntPoint(adjustment.width(), adjustment.height());
+        LOG_WITH_STREAM(Scrolling, stream << "ScrollAnchoringController::adjustScrollPositionForAnchoring() for frame: " << m_frameView << " adjusting from: " << frameView().scrollPosition() << " to: " << newScrollPosition);
```
This breaks Win/PS debug builds: Logging.h is missing, so `LOG_CHANNEL_PREFIXScrolling` is undeclared.

nmoucht replied:
Gotcha made https://github.com/WebKit/WebKit/pull/18062 to fix this.
